sled-agent/src/rack_setup/service.rs

```diff
@@ -51,176 +93,450 @@ impl Service {
+                        info!(
+                            self.log,
+                            "Waiting for {} peers (currently have {})",
+                            wanted_peer_count,
+                            all_addrs.len(),
+                        );
+                    }
+                }
+            }
+
+            info!(self.log, "Waiting for more peers");
+            let new_peer = peer_rx.recv().await?;
+            all_addrs.insert(new_peer);
+        }
     }
 
     // In lieu of having an operator send requests to all sleds via an
     // initialization service, the sled-agent configuration may allow for the
     // automated injection of setup requests from a sled.
+    //
+    // This method has a few distinct phases, identified by files in durable
+    // storage:
+    //
+    // 1. ALLOCATION PLAN CREATION. When the RSS starts up for the first time,
+    //    it creates an allocation plan to provision subnets and services
+    //    to an initial set of sleds.
+    //
+    //    This plan is stored at "rss_plan_path()".
+    //
+    // 2. ALLOCATION PLAN EXECUTION. The RSS then carries out this plan, making
+    //    requests to the sleds enumerated within the "allocation plan".
+    //
+    // 3. MARKING SETUP COMPLETE. Once the RSS has successfully initialized the
+    //    rack, the "rss_plan_path()" file is renamed to
+    //    "rss_completed_plan_path()". This indicates that the plan executed
+    //    successfully, and no work remains.
     async fn inject_rack_setup_requests(
         &self,
         config: &Config,
     ) -> Result<(), SetupServiceError> {
         info!(self.log, "Injecting RSS configuration: {:#?}", config);
 
-        let serialized_config = toml::Value::try_from(&config)
-            .expect("Cannot serialize configuration");
-        let config_str = toml::to_string(&serialized_config)
-            .expect("Cannot turn config to string");
+        // We expect this directory to exist - ensure that it does, before any
+        // subsequent operations which may write configs here.
+        tokio::fs::create_dir_all(omicron_common::OMICRON_CONFIG_PATH).await?;
 
-        // First, check if this request has previously been made.
-        //
-        // Normally, the rack setup service is run with a human-in-the-loop,
-        // but with this automated injection, we need a way to determine the
-        // (destructive) initialization has occurred.
+        // Check if a previous RSS plan has completed successfully.
         //
-        // We do this by storing the configuration at "rss_config_path"
-        // after successfully performing initialization.
-        let rss_config_path =
-            std::path::Path::new(omicron_common::OMICRON_CONFIG_PATH)
-                .join("config-rss.toml");
-        if rss_config_path.exists() {
+        // If it has, the system should be up-and-running.
+        let rss_completed_plan_path = rss_completed_plan_path();
```

Comment:
why use this variable instead of the function?

Reply:
meh, I could use either one. The function returns a new `PathBuf`, so this saves a minuscule amount of memory by only allocating one buffer.